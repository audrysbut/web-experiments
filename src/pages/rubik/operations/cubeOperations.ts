export type CubeData = number[];

export const initialize = (): CubeData => {
  return Array.from(Array(54).keys());
};

//TODO: move operations into separate files
export const left = (input: CubeData): CubeData => {
  const val = [...input];
  //TL
  const tl = val[45];
  val[45] = val[36];
  val[36] = val[18];
  val[18] = val[0];
  val[0] = tl;

  //ML
  const ml = val[48];
  val[48] = val[39];
  val[39] = val[21];
  val[21] = val[3];
  val[3] = ml;

  //BL
  const bl = val[51];
  val[51] = val[42];
  val[42] = val[24];
  val[24] = val[6];
  val[6] = bl;

  //rotate corners
  const rtl = val[9];
  val[9] = val[15];
  val[15] = val[17];
  val[17] = val[11];
  val[11] = rtl;

  //rotate center
  const rtm = val[10];
  val[10] = val[12];
  val[12] = val[16];
  val[16] = val[14];
  val[14] = rtm;

  return val;
};

export const leftB = (input: CubeData): CubeData => {
  //TODO: implement better
  return left(left(left(input)));
};

export const right = (input: CubeData): CubeData => {
  const val = [...input];

  //TR
  const tr = val[2];
  val[2] = val[20];
  val[20] = val[38];
  val[38] = val[47];
  val[47] = tr;

  //MR
  const mr = val[5];
  val[5] = val[23];
  val[23] = val[41];
  val[41] = val[50];
  val[50] = mr;

  //BR
  const br = val[8];
  val[8] = val[26];
  val[26] = val[44];
  val[44] = val[53];
  val[53] = br;

  //rotate corners
  const rtl = val[27];
  val[27] = val[33];
  val[33] = val[35];
  val[35] = val[29];
  val[29] = rtl;

  //rotate center
  const rtm = val[28];
  val[28] = val[30];
  val[30] = val[34];
  val[34] = val[32];
  val[32] = rtm;

  return val;
};

export const rightB = (input: CubeData) => {
  //TODO: implement better
  return right(right(right(input)));
};

export const top = (input: CubeData) => {
  const val = [...input];

  //rotate corner
  const rtl = val[0];
  val[0] = val[6];
  val[6] = val[8];
  val[8] = val[2];
  val[2] = rtl;

  //rotate center
  const rtm = val[1];
  val[1] = val[3];
  val[3] = val[7];
  val[7] = val[5];
  val[5] = rtm;

  //TL
  const tl = val[9];
  val[9] = val[18];
  val[18] = val[27];
  val[27] = val[53];
  val[53] = tl;

  //TM
  const tm = val[10];
  val[10] = val[19];
  val[19] = val[28];
  val[28] = val[52];
  val[52] = tm;

  //TR
  const tr = val[11];
  val[11] = val[20];
  val[20] = val[29];
  val[29] = val[51];
  val[51] = tr;

  return val;
};

export const topB = (input: CubeData) => {
  //TODO: implement better
  return top(top(top(input)));
};

export const bottom = (input: CubeData): CubeData => {
  const val = [...input];

  //rotate corner
  const rtl = val[36];
  val[36] = val[42];
  val[42] = val[44];
  val[44] = val[38];
  val[38] = rtl;

  //rotate center
  const rtm = val[37];
  val[37] = val[39];
  val[39] = val[43];
  val[43] = val[41];
  val[41] = rtm;

  //BR
  const br = val[35];
  val[35] = val[26];
  val[26] = val[15];
  val[15] = val[45];
  val[45] = br;

  //BM
  const bm = val[34];
  val[34] = val[25];
  val[25] = val[16];
  val[16] = val[46];
  val[46] = bm;

  //BL
  const bl = val[33];
  val[33] = val[24];
  val[24] = val[17];
  val[17] = val[47];
  val[47] = bl;

  return val;
};

export const bottomB = (input: CubeData) => {
  //TODO: implement better
  return bottom(bottom(bottom(input)));
};

export const back = (input: CubeData): CubeData => {
  const val = [...input];

  //rotate corner
  const rtl = val[45];
  val[45] = val[51];
  val[51] = val[53];
  val[53] = val[47];
  val[47] = rtl;

  //rotate center
  const rtm = val[46];
  val[46] = val[48];
  val[48] = val[52];
  val[52] = val[50];
  val[50] = rtm;

  //BR
  const br = val[44];
  val[44] = val[15];
  val[15] = val[0];
  val[0] = val[29];
  val[29] = br;

  //MR
  const mr = val[32];
  val[32] = val[43];
  val[43] = val[12];
  val[12] = val[1];
  val[1] = mr;

  //BR
  const br2 = val[35];
  val[35] = val[42];
  val[42] = val[9];
  val[9] = val[2];
  val[2] = br2;

  return val;
};

export const backB = (input: CubeData) => {
  //TODO: implement better
  return back(back(back(input)));
};

export const front = (input: CubeData): CubeData => {
  const val = [...input];

  //rotate corner
  const rtl = val[18];
  val[18] = val[24];
  val[24] = val[26];
  val[26] = val[20];
  val[20] = rtl;

  //rotate center
  const rtm = val[19];
  val[19] = val[21];
  val[21] = val[25];
  val[25] = val[23];
  val[23] = rtm;

  //BR
  const br = val[8];
  val[8] = val[11];
  val[11] = val[36];
  val[36] = val[33];
  val[33] = br;

  //BM
  const bm = val[7];
  val[7] = val[14];
  val[14] = val[37];
  val[37] = val[30];
  val[30] = bm;

  //BL
  const bl = val[6];
  val[6] = val[17];
  val[17] = val[38];
  val[38] = val[27];
  val[27] = bl;

  return val;
}

export const frontB = (input: CubeData) => {
  //TODO: implement better
  return front(front(front(input)));
};