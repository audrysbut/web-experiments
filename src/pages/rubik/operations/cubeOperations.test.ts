import { back, backB, bottom, bottomB, front, frontB, initialize, left, leftB, right, rightB, top, topB } from "./cubeOperations";

test("left operation should work", () => {
  const init = initialize();
  const after = left(init);
  expect(after).toMatchSnapshot();
});

test("leftB operation should work", () => {
  const init = initialize();
  const after = leftB(init);
  expect(after).toMatchSnapshot();
});

test("right operation should work", () => {
  const init = initialize();
  const after = right(init);
  expect(after).toMatchSnapshot();
});

test("rightB operation should work", () => {
  const init = initialize();
  const after = rightB(init);
  expect(after).toMatchSnapshot();
});

test("top operation should work", () => {
  const init = initialize();
  const after = top(init);
  expect(after).toMatchSnapshot();
});

test("topB operation should work", () => {
  const init = initialize();
  const after = topB(init);
  expect(after).toMatchSnapshot();
});

test("bottom operation should work", () => {
  const init = initialize();
  const after = bottom(init);
  expect(after).toMatchSnapshot();
});

test("bottomB operation should work", () => {
  const init = initialize();
  const after = bottomB(init);
  expect(after).toMatchSnapshot();
});

test("back operation should work", () => {
  const init = initialize();
  const after = back(init);
  expect(after).toMatchSnapshot();
});

test("backB operation should work", () => {
  const init = initialize();
  const after = backB(init);
  expect(after).toMatchSnapshot();
});

test("front operation should work", () => {
  const init = initialize();
  const after = front(init);
  expect(after).toMatchSnapshot();
});

test("frontB operation should work", () => {
  const init = initialize();
  const after = frontB(init);
  expect(after).toMatchSnapshot();
});