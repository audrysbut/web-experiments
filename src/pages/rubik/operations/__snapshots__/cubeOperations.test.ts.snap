// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`back operation should work 1`] = `
Array [
  29,
  32,
  35,
  3,
  4,
  5,
  6,
  7,
  8,
  2,
  10,
  11,
  1,
  13,
  14,
  0,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  44,
  30,
  31,
  43,
  33,
  34,
  42,
  36,
  37,
  38,
  39,
  40,
  41,
  9,
  12,
  15,
  51,
  48,
  45,
  52,
  49,
  46,
  53,
  50,
  47,
]
`;

exports[`backB operation should work 1`] = `
Array [
  15,
  12,
  9,
  3,
  4,
  5,
  6,
  7,
  8,
  42,
  10,
  11,
  43,
  13,
  14,
  44,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  0,
  30,
  31,
  1,
  33,
  34,
  2,
  36,
  37,
  38,
  39,
  40,
  41,
  35,
  32,
  29,
  47,
  50,
  53,
  46,
  49,
  52,
  45,
  48,
  51,
]
`;

exports[`bottom operation should work 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  45,
  46,
  47,
  18,
  19,
  20,
  21,
  22,
  23,
  17,
  16,
  15,
  27,
  28,
  29,
  30,
  31,
  32,
  24,
  25,
  26,
  42,
  39,
  36,
  43,
  40,
  37,
  44,
  41,
  38,
  35,
  34,
  33,
  48,
  49,
  50,
  51,
  52,
  53,
]
`;

exports[`bottomB operation should work 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  26,
  25,
  24,
  18,
  19,
  20,
  21,
  22,
  23,
  33,
  34,
  35,
  27,
  28,
  29,
  30,
  31,
  32,
  47,
  46,
  45,
  38,
  41,
  44,
  37,
  40,
  43,
  36,
  39,
  42,
  15,
  16,
  17,
  48,
  49,
  50,
  51,
  52,
  53,
]
`;

exports[`front operation should work 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  17,
  14,
  11,
  9,
  10,
  36,
  12,
  13,
  37,
  15,
  16,
  38,
  24,
  21,
  18,
  25,
  22,
  19,
  26,
  23,
  20,
  6,
  28,
  29,
  7,
  31,
  32,
  8,
  34,
  35,
  33,
  30,
  27,
  39,
  40,
  41,
  42,
  43,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  51,
  52,
  53,
]
`;

exports[`frontB operation should work 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  27,
  30,
  33,
  9,
  10,
  8,
  12,
  13,
  7,
  15,
  16,
  6,
  20,
  23,
  26,
  19,
  22,
  25,
  18,
  21,
  24,
  38,
  28,
  29,
  37,
  31,
  32,
  36,
  34,
  35,
  11,
  14,
  17,
  39,
  40,
  41,
  42,
  43,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  51,
  52,
  53,
]
`;

exports[`left operation should work 1`] = `
Array [
  45,
  1,
  2,
  48,
  4,
  5,
  51,
  7,
  8,
  15,
  12,
  9,
  16,
  13,
  10,
  17,
  14,
  11,
  0,
  19,
  20,
  3,
  22,
  23,
  6,
  25,
  26,
  27,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  18,
  37,
  38,
  21,
  40,
  41,
  24,
  43,
  44,
  36,
  46,
  47,
  39,
  49,
  50,
  42,
  52,
  53,
]
`;

exports[`leftB operation should work 1`] = `
Array [
  18,
  1,
  2,
  21,
  4,
  5,
  24,
  7,
  8,
  11,
  14,
  17,
  10,
  13,
  16,
  9,
  12,
  15,
  36,
  19,
  20,
  39,
  22,
  23,
  42,
  25,
  26,
  27,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  45,
  37,
  38,
  48,
  40,
  41,
  51,
  43,
  44,
  0,
  46,
  47,
  3,
  49,
  50,
  6,
  52,
  53,
]
`;

exports[`right operation should work 1`] = `
Array [
  0,
  1,
  20,
  3,
  4,
  23,
  6,
  7,
  26,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  38,
  21,
  22,
  41,
  24,
  25,
  44,
  33,
  30,
  27,
  34,
  31,
  28,
  35,
  32,
  29,
  36,
  37,
  47,
  39,
  40,
  50,
  42,
  43,
  53,
  45,
  46,
  2,
  48,
  49,
  5,
  51,
  52,
  8,
]
`;

exports[`rightB operation should work 1`] = `
Array [
  0,
  1,
  47,
  3,
  4,
  50,
  6,
  7,
  53,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  2,
  21,
  22,
  5,
  24,
  25,
  8,
  29,
  32,
  35,
  28,
  31,
  34,
  27,
  30,
  33,
  36,
  37,
  20,
  39,
  40,
  23,
  42,
  43,
  26,
  45,
  46,
  38,
  48,
  49,
  41,
  51,
  52,
  44,
]
`;

exports[`top operation should work 1`] = `
Array [
  6,
  3,
  0,
  7,
  4,
  1,
  8,
  5,
  2,
  18,
  19,
  20,
  12,
  13,
  14,
  15,
  16,
  17,
  27,
  28,
  29,
  21,
  22,
  23,
  24,
  25,
  26,
  53,
  52,
  51,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  38,
  39,
  40,
  41,
  42,
  43,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  11,
  10,
  9,
]
`;

exports[`topB operation should work 1`] = `
Array [
  2,
  5,
  8,
  1,
  4,
  7,
  0,
  3,
  6,
  53,
  52,
  51,
  12,
  13,
  14,
  15,
  16,
  17,
  9,
  10,
  11,
  21,
  22,
  23,
  24,
  25,
  26,
  18,
  19,
  20,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  38,
  39,
  40,
  41,
  42,
  43,
  44,
  45,
  46,
  47,
  48,
  49,
  50,
  29,
  28,
  27,
]
`;
